packages/core/fs/src/MemoryFS.js

```
@@ -0,0 +1,493 @@
+          this.files.set(destName, buffer);
+        }
+      }
+    } else {
+      await this.copyFile(source, destination);
+    }
+  }
+
+  createReadStream(filePath: FilePath) {
+    return new ReadStream(this, filePath);
+  }
+
+  createWriteStream(filePath: FilePath, options: ?FileOptions) {
+    return new WriteStream(this, filePath, options);
+  }
+
+  async realpath(filePath: FilePath) {
+    filePath = this._normalizePath(filePath);
+    return filePath;
+  }
+
+  async exists(filePath: FilePath) {
+    filePath = this._normalizePath(filePath);
+    return this.files.has(filePath) || this.dirs.has(filePath);
+  }
+}
+
+class ReadStream extends Readable {
+  fs: FileSystem;
+  filePath: FilePath;
+  reading: boolean;
+  bytesRead: number;
+  constructor(fs: FileSystem, filePath: FilePath) {
+    super();
+    this.fs = fs;
+    this.filePath = filePath;
+    this.reading = false;
+    this.bytesRead = 0;
+  }
+
+  _read() {
+    if (this.reading) {
+      return;
+    }
+
+    this.reading = true;
+    this.fs.readFile(this.filePath).then(
+      res => {
+        this.bytesRead += res.byteLength;
+        this.push(res);
+        this.push(null);
+      },
+      err => {
+        this.emit('error', err);
+      }
+    );
+  }
+}
+
+class WriteStream extends Writable {
+  fs: FileSystem;
+  filePath: FilePath;
+  options: ?FileOptions;
+  buffer: Buffer;
+  constructor(fs: FileSystem, filePath: FilePath, options: ?FileOptions) {
+    super();
+    this.fs = fs;
+    this.filePath = filePath;
+    this.options = options;
+    this.buffer = Buffer.alloc(0);
+  }
+
+  _write(
+    chunk: Buffer | string,
+    encoding: any,
+    callback: (error?: Error) => void
+  ) {
+    let c = Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk, encoding);
+    this.buffer = Buffer.concat([this.buffer, c]);
+    callback();
+  }
+
+  _final(callback: (error?: Error) => void) {
+    this.fs.writeFile(this.filePath, this.buffer, this.options).then(callback);
```

Should rather be something like:
```js
.then(() => callback(), (err) => callback(err))
```
otherwise errors from `writeFile` would be silently swallowed

---

Good catch